ast/compile_test.go

```diff
@@ -109,6 +110,124 @@ func TestCompilerExample(t *testing.T) {
 	assertNotFailed(t, c)
 }
 
+func TestCompilerUserFunctionConflicts(t *testing.T) {
+	tests := []struct {
+		module string
+		assert func(*testing.T, *Compiler)
+	}{
+		{
+			// Test that functions can have different input types.
+			module: `package x
+
+				f([x]) = y {
+					y = x
+				}
+
+				f({"foo": x}) = y {
+					y = x
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f([x]) = y {
+					y = x
+				}
+
+				f([[x]]) = y {
+					y = x
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f(1, x) = y {
+					y = x
+				}
+
+				f(x, y) = z {
+					z = x+y
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f(x, 1) = y {
+					y = x
+				}
+
+				f(x, [y]) = z {
+					z = x+y
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f({"foo": {"bar": x}}) = y {
+					y = x
+				}
+
+				f({"foo": [x]}) = y {
+					y = x
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f(1) = y {
+					y = "foo"
+				}
+
+				f(2) = y {
+					y = 2
+				}`,
+			assert: assertNotFailed,
+		},
+		{
+			module: `package x
+
+				f(1) = y {
+					y = "foo"
+				}
+
+				f(2) = y {
+					y = "bar"
+				}`,
+			assert: assertNotFailed,
+		},
+		// Test that functions must have the same number of inputs.
```

Comment:
I remember having a discussion with @timothyhinrichs when we decided that requiring a consistent number of arguments would be reasonable.